Firestore/core/src/util/executor_libdispatch.mm

```diff
@@ -243,91 +94,231 @@ void MarkDone() {
+      // additional retain is required.
+      tasks_.erase(found_task);
       schedule_.erase(found);
     }
-  });
+  }
+
+  if (found_task) {
+    found_task->Cancel();
+
+    // Release this method's ownership.
+    found_task->Release();
+  }
+}
+
+void ExecutorLibdispatch::InvokeAsync(void* raw_task) {
+  auto* task = static_cast<Task*>(raw_task);
+  task->ExecuteAndRelease();
+}
+
+void ExecutorLibdispatch::InvokeSync(void* raw_task) {
+  // Note: keep this implementation separate from `InvokeAsync` to make it
+  // clearer in stack traces exactly what's going on.
+  auto* task = static_cast<Task*>(raw_task);
+  task->ExecuteAndRelease();
 }
 
 // Test-only methods
 
-bool ExecutorLibdispatch::IsScheduled(const Tag tag) const {
-  bool result = false;
-  RunSynchronized(this, [this, tag, &result] {
-    result = std::any_of(schedule_.begin(), schedule_.end(),
-                         [&tag](const ScheduleEntry& operation) {
-                           return *operation.second == tag;
-                         });
-  });
-  return result;
+bool ExecutorLibdispatch::IsTagScheduled(Tag tag) const {
+  std::vector<Task*> matches;
+
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+
+    // There's a race inherent in making `IsTagScheduled` checks after a task
+    // has executed. The problem is that the task has to lock the Executor to
+    // report its completion, but `IsTagScheduled` needs that lock too. Absent
+    // any intervention, if `IsTagScheduled` won the race it could indicate
+    // that tasks that appear completed are still scheduled.
+    //
+    // Work around this by waiting for tasks that are currently executing to
+    // complete. That is, only tasks that are in the schedule and in the
+    // `kInitial` state are considered scheduled.
+    //
+    // Unfortunately, this has to be done with the Executor unlocked so that
+    // the task can report its completion without deadlocking. The executor
+    // can't be unlocked while iterating the `schedule_` though, so collect up
+    // potential matches in a separate collection.
+    for (const ScheduleEntry& entry : schedule_) {
+      Task* task = entry.second;
+      if (task->tag() == tag) {
+        matches.push_back(task);
+
+        // Retain local references to prevent the task from deleting itself
+        // before it can be examined outside the executor mutex.
+        task->Retain();
+      }
+    }
+  }
+
+  bool tag_scheduled = false;
+  for (Task* task : matches) {
+    bool task_completed = task->AwaitIfRunning();
+    tag_scheduled = tag_scheduled || !task_completed;
+
+    // Release this method's ownership.
+    task->Release();
+  }
+
+  return tag_scheduled;
 }
 
-absl::optional<Executor::TaggedOperation>
-ExecutorLibdispatch::PopFromSchedule() {
-  absl::optional<Executor::TaggedOperation> result;
+bool ExecutorLibdispatch::IsIdScheduled(Id id) const {
+  Task* found_task = nullptr;
```

Comment:
Looking at these functions, it seems like `Schedule` is a clearly "separatable" abstraction (similar to `ExecutorStd`) that could wrap the mutex and the two containers and potentially return `TaskRef`s. It might be overkill, though.

Reply:
I thought about this too. For now I've left it as is just to avoid further iterating on this PR.